server/chatroom.py

```diff
@@ -35,21 +46,104 @@ def saveMessage(cur, sender_id, chatroomID, message_sent):
             {"sender_id": sender_id, "chatroomID": chatroomID, "message_sent": message_sent,
              "date_time_sent": date_time_sent})
         return True
-    except:
-        return False
+    except BaseException as e:
+        print(f'Exception: {e}')
+        raise HTTPException(
+            status_code=500,
+            detail="Database Error",
+        )
 
 
 # Create non-course chatroom
 def createChatroom(cur, user_id, chatroom_name, uni_id):
     try:
-        cur.execute("INSERT INTO chatroom(chatroom_name, uni_id) VALUES(%(chatroom_name)s, %(uni_id)s)",
-                    {"chatroom_name": chatroom_name, "uni_id": uni_id})
-        cur.execute("SELECT id FROM chatroom where chatroom_name = %(chatroom_name)s   AND uni_id = %(uni_id)s",
-                    {"chatroom_name": chatroom_name, "uni_id": uni_id})
+        # Generate invite id
+        invite_id = ''.join(random.choices(string.ascii_uppercase + string.digits, k=10))
+        
+        cur.execute("INSERT INTO chatroom(chatroom_name, uni_id, invite_id) VALUES(%(chatroom_name)s, %(uni_id)s, %(invite_id)s)", {"chatroom_name" : chatroom_name, "uni_id" : uni_id, "invite_id" : invite_id})
+        
+        cur.execute("SELECT id FROM chatroom where chatroom_name = %(chatroom_name)s   AND uni_id = %(uni_id)s", {"chatroom_name" : chatroom_name, "uni_id" : uni_id})
+        
         chatroom_id = cur.fetchall()[0][0]
+        
         cur.execute(
             "INSERT INTO in_chatroom(student_id, uni_id, chatroom_id) VALUES(%(user_id)s, %(uni_id)s, %(chatroom_id)s)",
             {"user_id": user_id, "uni_id": uni_id, "chatroom_id": chatroom_id})
+        
         return True
-    except:
-        return False
+    except BaseException as e:
+        print(f'Exception: {e}')
+        raise HTTPException(
+            status_code=500,
+            detail="Database Error",
+        )
+    
+#Generate invite link for student (for non-course chatroom)
+def GenerateInvite(cur, target_user_id, chatroom_id):
+    try:
+        invite_object = {}
+        invite_object.update({"target_user" : target_user_id})
+        cur.execute("SELECT invite_id FROM chatroom WHERE id = %(chatroom_id)s", {"chatroom_id" : chatroom_id})
+        invite_id = cur.fetchall()[0][0]
+        invite_object.update({"invite_id" : invite_id})
+        return json.dumps(invite_object)
+    except BaseException as e:
+        print(f'Exception: {e}')
+        raise HTTPException(
+            status_code=500,
+            detail="Database Error",
+        )
+
+#Accept an invite and join a (non-course) chatroom
+def AcceptInvite(cur, invite_object, target_user_id):
+    try:
+        invite = json.loads(invite_object)
```

Comment:
what is stopping the user to make up their own invite object? how can we check this? perhaps we should store the invite objects in the db (like an invitation table)?

Reply:
it's likely still insufficient. what we can do is make an invitation table. we store that invite code. and when user accepts the invitation, the user will give us the code, and we will just get the chatroom id from the table instead of letting users pass their own. lastly delete the entry on invitation when we're done